Add: nox config for docs, tests

lwasser opened the pull request:
closes #395

<!--- * If this pr does not address an open issue  in the repository, please be sure
to explain what this pull request fixes or does. We prefer to discuss changes in issues first, to just ensure we are all on the same page about the change being proposed.

* If this is a technical code change, please be sure that you have [read our contributing guide.](https://stravalib.readthedocs.io/en/latest/contributing/how-to-contribute.html)
-->

## Description

This change adds Nox to our build workflow. Right now we are using Make for everything. Nox is much cleaner because it's python and as such easy to read. it also supports virtual environments and running things across various versions for local debugging. 

This pr does NOT include building our package but that step will be easy to add. I just wanted folks to test it out to see if everyone is comfortable with moving to nox and removing the old-school Makefile that is much harder to read / maintain (IMO) :) with less fun features like environments and such for robust test dev locally.  ✨ 


## Type of change

Select the statement best describes this pull request.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] This is a documentation update
- [X] Other (please describe) 

This is an infrastructure update!

## Does your PR include tests

If you are fixing a bug or adding a feature, we appreciate (but do not require)
tests to support whatever fix of feature you're implementing.

- [ ] Yes
- [ ] No, i'd like some help with tests
- [x] This change doesn't require tests

This helps us run our tests!

## Did you include your contribution to the change log?

- [ ] Yes, `changelog.md` is up-to-date.

*If you are having a hard time getting the tests to run correctly feel free to
ping one of the maintainers here!*

There is no change log yet, if folks like this approach i'll 

- [ ] add a build step
- [ ] remove make from our workflow
- [ ] update our documentation 

i was inspired to do this given the current issues with python 3.8. it's just easier to be able to run test locally on the version that i want without switching environments all of the time. 

test this out now!!

try
``` 
pip install nox
nox -s docs-live
```

docs-live session will start an interactive session where as you edit documentation files, the docs themselves will update in your browser. NOTE: i've had issues with api docs not updating but everything else should update!

```
nox -s test
```
if you use venv for environments this is for you. run tests on python 3.8 thru 3.11 locally!!
NOTE: you will need to install the various versions of python locally for this to work on your machine!

or if you are a mamba fan like me:

```
nox -s test_mamba
```

this will use conda/mamba environments!

enadeau commented:
I've been trying to play with this but I'm getting 
```
nox > Command python -m pip install -r requirements-test.txt failed with exit code 1:
ERROR: Could not open requirements file: [Errno 2] No such file or directory: 'requirements-test.txt'
```
Looks like you forgot to push a file

lwasser commented on noxfile.py:
```diff
@@ -0,0 +1,75 @@
+@nox.session(python=["3.8", "3.9", "3.10", "3.11"])
+def test(session):
+    session.install(".[all]")
+    session.install("-r", "requirements-test.txt")
```
```suggestion
    session.install("-r", "requirements.txt")
```

lwasser commented on noxfile.py:
```diff
@@ -0,0 +1,75 @@
+@nox.session(venv_backend="mamba", python=["3.8", "3.9", "3.10", "3.11"])
+def test_mamba(session):
+    session.install(".[all]")
+    session.install("-r", "requirements-test.txt")
```
```suggestion
    session.install("-r", "requirements.txt")
```

lwasser replied:
@enadeau just so you can see - i was playing with various requirements files and definitely made a mistake in not updating these functions!! each one of these functions defines a session.

you can create whatever type of session you want. for instance the docs session allows us to build the docs and they will update in realtime as we work on them! If you just run nox it will run all sessions - the way i have things here that wouldn't be ideal as there is a conda and a venv tests option.

lwasser commented:
oops - fixed. that was not a forget to push, but rather a blatant mistake that i didn't mean to push to this pr :)  great catch! it's fixed now and i left you a comment so you can see how the sessions are created! it's all python which is why i really like it. i'm not sure if tox allows you to define other custom local "sessions" like nox does? but nox is really flexible and is becoming a go to local build tool for me. i'll be curious to see how you like it!

lwasser commented:
one other note - i was just thinking about this @enadeau i think the best fix for that test run would be to add a src/ directory to stravalib. so if we move forward with this we could simplify our builds using

1.  src/stravalib directory
2. using nox to build and test across versions locally
3. removing make
4. adding a  sphinx autobuild so we can serve docs and see realtime edits locally. 

i think that would be a great infrastructure upgrade for us here. but want @jsamoocha and @yihong0618 to weight in before making such changes. 

if we wanted to iterate i could move to src/ first and not move to nox as a baby step.

enadeau commented:
Very nice it works well for me now!

I like the idea of moving away from make and use nox instead.

I think it would be nice to also have a session to run mypy :smiley:

yihong0618 reviewed (approved):
LGTM

lwasser commented:
ok - yes we could add mypy and any other sessions we wish to. 

i think we could (for now) setup everything with nox, i can update our docs but leave make just until everyone is comfortable with moving away from make. 

how about i add a bit more dev documentation to this pr on using nox. and i'll give mypy a shot locally with it too to see how it goes. we could even run it across python versions if we wish to. 

Then maybe in another PR we could consider moving our project structure to src/stravalib which would then make local builds always test on the installed version.

enadeau reviewed (approved):
LGTM. Will you be tackling the documentation update in an other PR? IF yes then you can merge I think

lwasser commented:
how about this - i'll add some docs in the next day or two just for this set of updates and then we can merge!

enadeau commented:
Great let's wait to merge then

lwasser commented on docs/contributing/development-guide.md:
````diff
@@ -174,22 +179,39 @@ force the branch to be in the same commit state as your local branch.
 
 ```
 
+### Typing using mypy
````
@enadeau i added this so you can test running mypy locally across python versions now!! :)

lwasser commented on noxfile.py:
```diff
@@ -0,0 +1,91 @@
+
+# Use this for venv envs nox -s test
+@nox.session(python=["3.8", "3.9", "3.10", "3.11"])
+def mypy(session):
```
this is the mypy session. when we drop python 3.8 we can drop it here too and in the test runs.

lwasser commented:
Ok! this is ready for another review! i
1. added a mypy session so we can run mypy across all python versions. 
2. updated docs
3. replaced master with main when i updated 

i didn't go into too much detail about how nox works but could add some things if people start using it and want a bit more info!! let me know if you catch anything that is missing, wrong, etc!

i figure in another pr i can move stravalib to a src/ layout and update nox to build the package.

enadeau commented:
It looks good to me but there seems to be a couple of mention of `make` left in the doc. Is it that some operation are till only in the makefile? Otherwise maybe a good crtl-f could clean that.

Agree that we should keep the structure change for another PR.

enadeau commented:
Maybe also best to remove 3.8 right now from the nox file since #416 will remove support for 3.8

lwasser commented:
ok i've added a few more fixes here

* removed 3.8 from nox build
* caught a few more references to make -C docs
* fixed some small other typos. 

let me know if you catch anything else!

lwasser commented:
Ok - 

✔️ docs are now building again
✔️ documentation is added for nox
✔️ tried to remove legacy references to master branch in docs

I think this can be merged now. if no one sees any other issues i can merge tomorrow!!

lwasser commented:
ok merging this now!! we can always modify / edit docs as needed.
